Subject: Defensive to encoding of output in FileConsumer
 * Add '>:encoding(EncodeTo)' line discipline to FileConsumer to defeat 
 perl's automatic charset conversion.
Author: Florian Ragwitz <rafl@debian.org>
Reviewed-by: Nicholas Bamber <nicholas@periapt.co.uk>
Forwarded: no
Last-Update: 2010-08-22
--- a/lib/XML/SAX/Writer.pm
+++ b/lib/XML/SAX/Writer.pm
@@ -330,7 +330,10 @@
 # new
 #-------------------------------------------------------------------#
 sub new {
-    my ( $proto, $file ) = ( shift, shift );
+    my ( $proto, $file, $opt ) = @_;
+    my $enc_to = (defined $opt and ref $opt eq 'HASH'
+                  and defined $opt->{EncodeTo}) ? $opt->{EncodeTo}
+                                                : 'utf-8';
 
     XML::SAX::Writer::Exception->throw(
         Message => "No filename provided to " . ref( $proto || $proto )
@@ -338,9 +341,10 @@
 
     local *XFH;
 
-    open XFH, ">$file" or XML::SAX::Writer::Exception->throw(
+    open XFH, ">:encoding($enc_to)", $file
+      or XML::SAX::Writer::Exception->throw(
         Message => "Error opening file $file: $!"
-    );
+      );
 
     return $proto->SUPER::new( *{XFH}{IO}, @_ );
 }
